Boost: Introduce base minify class for JS and CSS modules

dilirity opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Add a base Minify class that handles deactivation of the minification modules. If both modules (JS and CSS) are deactivated this module will remove the WP Cron events they set up.

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Add a Minify class. The activate function sets up the cron jobs required. The deactivate function removes them, if both CSS and JS modules are deactivated.
* Map the JS and CSS modules to the Minify module.
* When setting the module status, set it in the Minify module depending on if either the css or js module are active.
* Add the Minify class to the Modules Index.
* When a module's status changes, and it's the css or js concatenate one, then deactivate the Minify one if both are deactivated, or if activating either of them, activate the Minify module.


### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Apply patch
* Add WP Cron Control plugin
* Toggle both Concatenation modules on.
* Toggle one off
* Confirm the jetpack_boost_404_tester_cron and jetpack_boost_minify_cron_cache_cleanup jobs are still there.
* Toggle the other module off.
* The two WP Cron jobs should be gone now.

donnchawp committed:
Add function that unschedules the 404 tester

donnchawp committed:
changelog

donnchawp committed:
Clear the cache cleanup cron too

donnchawp committed:
Clear the 404 tester cron on plugin deactivation too

donnchawp committed:
Schedule the daily jobs from the setup, rather than activation

If the plugin is deactivated with the concatenation modules enabled
then the activate function isn't called when the plugin is activated
again.
The setup function runs on every pageload, but it does an is_admin()
check to reduce DB queries.

donnchawp committed:
Revert changes to setup function

donnchawp committed:
Use base Minify class for both modules

donnchawp committed:
This code doesn't belong here

donnchawp committed:
Revert previous changes and rename Minify php file

donnchawp committed:
Track how many minify modules are active in Minify

donnchawp committed:
Merge branch 'trunk' into update/boost/unified-minify-class

donnchawp committed:
Changelog

donnchawp committed:
Remove duplicate deactivation function

donnchawp committed:
Call Minify activate and deactivate from JS and CSS modules

donnchawp committed:
Merge remote-tracking branch 'origin/trunk' into update/boost/unified-minify-class

donnchawp committed:
Move these functions back to where they were before

donnchawp committed:
move get_slug around too

donnchawp committed:
Activate the Minify module by mapping js/css to it

donnchawp committed:
Schedule before testing as scheduling fails.

I'm not sure why, but if I schedule the job before running the test, the
schedule fails.

donnchawp committed:
No need for this, it's called by activation.

The 404 tester is called when it's scheduled

donnchawp committed:
Schedule the tester, and let cron run the first test immediately

donnchawp committed:
Module must be "activated" to call the on_activate function

Without this change, the on_activate() function would be called when one
module was toggled off, but the other was on, since the previous
! jetpack_boost_minify_is_enabled() check would be false.
We only want on_deactivate() to be called when both are disabled, but
on_activate() must be called whenever either module is activated.

donnchawp committed:
And revert back again. Need the tester for the UI

donnchawp committed:
Restore this blank line.

LiamSarsfield commented on projects/plugins/boost/app/modules/Modules_Setup.php:
```diff
@@ -149,6 +149,18 @@ public function on_module_status_update( $module_slug, $is_activated ) {
 		}
 
+		// The Minify module handles de/activation of the CSS and JS modules.
+		if ( in_array( $module_slug, array( 'minify_css', 'minify_js' ), true ) ) {
```
Just thinking out loud; I wonder could we do this in a way that doesn't explicitly check specific modules (In this case Minify CSS/JS) in the Module_Setup?  It feels like the Minify class should be the one calling the [on_activate](https://github.com/Automattic/jetpack/pull/41716/files#diff-d43d5780514f1d715c5bbee67e66f52fb5d93cd10e0e03641945a5d2fb3b006bR145) / [on_deactivate](https://github.com/Automattic/jetpack/pull/41716/files#diff-d43d5780514f1d715c5bbee67e66f52fb5d93cd10e0e03641945a5d2fb3b006bR149) here, and the Minify class calling either the Minify CSS/JS. 
This would follow the "Tell, Don't Ask" principle, where instead of checking what type of module something is, we let the module handle its own behavior.
I could be wrong here, but it feels like a code smell having explicit checks for specific slugs in a generic class like this.
What do you think about this approach? Happy to discuss further!

donnchawp replied:
Agreed. As per Slack conversation, my first idea for this was to use a base class that the others inherit from, but I think I can use sub modules. Make the Minify module a sub module that does the final deactivation when both css and js modules are deactivated.

Thanks!

LiamSarsfield commented on projects/plugins/boost/app/lib/Status.php:
```diff
@@ -38,6 +38,12 @@ public function __construct( $slug ) {
 			Cloud_CSS::get_slug() => array(
 				Critical_CSS::get_slug(),
 			),
+			'minify_css'          => array(
+				'minify',
+			),
+			'minify_js'           => array(
+				'minify',
+			),
 		);
 	}
```
Could we use the slug from the class here?
```suggestion
			Cloud_CSS::get_slug()  => array(
				Critical_CSS::get_slug(),
			),
			Minify_CSS::get_slug() => array(
				Minify::get_slug(),
			),
			Minify_JS::get_slug()  => array(
				Minify::get_slug(),
			),
		);
	}
```

donnchawp replied:
Good idea!

dilirity replied:
I just noticed that the suggestion has an additional line that caused the PHP error to happen. It's also missing the use statements for the classes.

LiamSarsfield reviewed (commented):
Per testing instructions, works for me 🚀 If my suggestion/comment is not necessary, happy to approve 🙌

donnchawp committed:
Update projects/plugins/boost/app/lib/Status.php

Use the get_slug() method to return the slug.

Co-authored-by: Liam Sarsfield <43409125+LiamSarsfield@users.noreply.github.com>

donnchawp committed:
Fix fatal errors from last code checked in.

dilirity reviewed (commented):
Since I created this PR, I can't approve, but I'll leave a comment. We discussed this in the thread, but I want to have it here as well.

I'm not sure how I feel about having the `Minify` class in the list of features. I kind of feel like it's just a short-cut to reduce code repetition. It also introduces a "feature" that does technically nothing.

donnchawp committed:
Use sub modules for the Minify module

This will allow this code to remain generic and still deactivate the
concatenation modules correctly

donnchawp committed:
Make sure it's an array before using the sub modules.

donnchawp committed:
Merge branch 'trunk' into update/boost/unified-minify-class

donnchawp committed:
Call the tester after, so it doesn't break scheduling

donnchawp committed:
Use get_available_submodules() to get the submodules

LiamSarsfield commented on projects/plugins/boost/app/modules/optimizations/minify/class-minify-css.php:
```diff
@@ -68,12 +72,22 @@ public function init_minify() {
+	/**
+	 * This is called only when the module is activated.
+	 */
 	public static function activate() {
```
Nit: Can we remove the `activate` function here and remove the `Has_Activate` interface?
Also for the JS Minify module

donnchawp replied:
Well spotted. PR updated, I'll merge it once the checks pass. Thanks!

donnchawp committed:
Remove the activate functions, they're not used here

LiamSarsfield reviewed (dismissed):
LGTM, works as expected and happy to approve based on offline discussion p1740054515101419-slack-C016BBAFHHS
